### PR TITLE
TYP: enable reportUnusedImport

### PR DESCRIPTION
**Pull request by twoertwein**
pyright's reportUnusedImport checks py and pyi files for unused imports. If an import is explicitly marked as public (re-exported using `as` or in `__all__`) it is "used" (flake8 seems to use the same definition).

Adding unused imports to `__all__` becomes messy when the file does not yet have `__all__`: need to list also all public symbols (constants, classes, functions).

### CODE REVIEW

File: pandas/tseries/offsets.py [hunk 1]
```diff
@@ -73,6 +78,7 @@
     "Week",
     "WeekOfMonth",
     "Easter",
+    "Tick",
```
Comment (twoertwein) [inline pandas/tseries/offsets.py:81]: When I add `Tick` to `__all__`,  `pytest -k test_sub_n_gt_1_ticks  pandas/tests/arithmetic/test_period.py` fails with:

```py
cls = <class 'pandas.core.dtypes.dtypes.PeriodDtype'>, freq = <4 * Ticks>

    def __new__(cls, freq=None):
        """
        Parameters
        ----------
        freq : frequency
        """
        if isinstance(freq, PeriodDtype):
            return freq

        elif freq is None:
            # empty constructor for pickle compat
            # -10_000 corresponds to PeriodDtypeCode.UNDEFINED
            u = dtypes.PeriodDtypeBase.__new__(cls, -10_000)
            u._freq = None
            return u

        if not isinstance(freq, BaseOffset):
            freq = cls._parse_dtype_strict(freq)

        try:
            return cls._cache_dtypes[freq.freqstr]
        except KeyError:
>           dtype_code = freq._period_dtype_code
E           AttributeError: 'pandas._libs.tslibs.offsets.Tick' object has no attribute '_period_dtype_code'

pandas/core/dtypes/dtypes.py:863: AttributeError
```

I think the only runtime effect this should have is on `from pandas.tseries.offset import *` Why is `Tick` (and the other unused imports) imported in this file? I assume they are imported for some side effect?

@jbrockmendel you made changes in #34227 which might help explain this weird behavior

Comment (jbrockmendel) [reply pandas/tseries/offsets.py:81]: This is bc Tick never gets instantiated directly, just Tick subclasses

Comment (twoertwein) [reply pandas/tseries/offsets.py:81]: Thanks, it seems that when I add Tick to `__all__`, `Tick` is somehow added to `tick_classes` which causes the `Tick` classes to be tested itself.

Comment (twoertwein) [reply pandas/tseries/offsets.py:81]: `tick_classes` is defined here
https://github.com/pandas-dev/pandas/blob/a853022ea154dd38dd759300ee50b456f3a9ddf6/pandas/tests/tseries/offsets/test_ticks.py#L40
but I cannot see where `Tick` is added to the list. Is there a good way to debug fixtures with pytest?

Even though `Tick` cannot/shouldn't be instantiated, I think it should be public for users to annotate functions accepting any Tick-subclass.

Comment (twoertwein) [reply pandas/tseries/offsets.py:81]: found it :) the fixture is defined in https://github.com/pandas-dev/pandas/blob/a853022ea154dd38dd759300ee50b456f3a9ddf6/pandas/conftest.py#L1690

**Review by mroeschke — APPROVED** [review]
So just to confirm:

1. Nothing changed from the "public" API?
2. In terms of code flow, we will need to ensure all "unused" imports are defined in `__all__`?

**Comment by twoertwein** [comment]
`__all__` as two functions: 1) it affects `from xyz import *` (imports only functions/classes inside `__all__`) and 2) in a py.typed library `__all__` defines the public interface of a module (unless the module is private, starts with `_`).

>     1. Nothing changed from the "public" API?

Technically, users will import fewer modules if they use `from xyz import *` which could be a public API change.

>     2. In terms of code flow, we will need to ensure all "unused" imports are defined in `__all__`?

flake8/pyright will let you know about unused imports :) They can then either be removed, be ignored (if they have some side effects), or added to `__all__` if they are meant to be public.

**Comment by mroeschke** [comment]
> Technically, users will import fewer modules if they use from xyz import * which could be a public API change.

Could you do a quick before/after check when `from pandas import *` is run? I imagine it should hopefully not change in a negative way, and maybe can be noted as a benefit that less stuff is imported :)

**Comment by twoertwein** [comment]
Seems there is actually no differences between main and this PR when imported everything from a file:

```py
import importlib
from pathlib import Path
import pickle

import ipdb

imported = {}
for py in Path("pandas").glob("**/*.py"):
    print(py)
    py = str(py).replace("/", ".").removesuffix(".py")
    mod = importlib.import_module("py")
    imported[py] = [x for x in dir(mod) if not x.startswith("_")]
with open("before.pickle", mode="wb") as file:
    pickle.dump(imported, file)

# ran the above first on this PR and save as after.pickle
# and then the entire script on main

after = {}
with open("after.pickle", mode="rb") as file:
    after = pickle.load(file)

for py, imports_b in imported.items():
    imports_a = set(after[py])
    imports_b = set(imports_b)
    remove = imports_b.difference(imports_a)
    added = imports_a.difference(imports_b)

    if len(remove) > 0 or len(added) > 0:
        ipdb.set_trace()  # never ended up here
```

**Comment by mroeschke** [comment]
Thanks for double checking @twoertwein!

**Pull request merged**